Add physical cluster replication tutorials and guides (#17915)
cockroachdb/docs: src/current/v23.2/set-up-physical-cluster-replication.md

    ~~~
    id |    name     | data_state | service_mode
    ---+-------------+------------+---------------
    1  | system      | ready      | shared
    2  | template    | ready      | none
    3  | application | ready      | shared
    (3 rows)
    ~~~

### Create a replication user and password

The standby cluster requires a user profile on the primary cluster to connect to.

1. From the primary's system interface SQL shell, create a user and password:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE USER {your username} WITH LOGIN PASSWORD '{your password}';
    ~~~

1. Grant the [`REPLICATION` system privilege]({% link {{ page.version.version }}/security-reference/authorization.md %}#supported-privileges) to your user: {% comment %}Link to detail on replication system privilege{% endcomment %}

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    GRANT SYSTEM REPLICATION TO {your username};
    ~~~

    If you need to change the password later, refer to [`ALTER USER`]({% link {{ page.version.version }}/alter-user.md %}).

### Connect to the primary virtual cluster (optional)

> 1. If you would like to add a workload to the primary's application virtual cluster, open a new terminal window, and connect with:

Reviewer: Not sure on the "add a workload" language. I think the value here is showing the user how to connect to the application tenant. They can use the ccluster=application option as you did below. But, if they've followed the instructions to this point, they can also elide that option as the application cluster will be the default.

Author: I have adjusted this slightly so that users know that the `ccluster=application` param will connect them to the application VC.

1. Place `ca_primary.crt` in the `certs` directory of the standby cluster nodes.

You need to add the `ca.crt` from the standby cluster to the `certs` directory on all the nodes in the primary cluster.

1. Name the `ca.crt` from the standby cluster to a new name on the primary cluster. For example, `ca_standby.crt`.
1. Place `ca_standby.crt` in the `certs` directory of the primary cluster nodes.

## Step 4. Start replication

The system interface in the standby cluster initiates and controls the replication stream by pulling from the primary cluster. In this section, you will connect to the primary from the standby to initiate the replication stream.

1. From the **standby** cluster, use your connection string to the primary.

    The connection string contains:
    - The replication user and password that you [created for the primary cluster](#create-a-replication-user-and-password).
    - The node IP address or hostname of one node from the primary cluster.
    - The path to the primary node's certificate on the standby cluster.

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE VIRTUAL CLUSTER standbyapplication

>     FROM REPLICATION OF application

Reviewer: I believe we'll want the user to also specify the template:

```
CREATE VIRTUAL CLUSTER standbyapplication LIKE template
FROM REPLICATION...`

Author: @stevendanna The `LIKE template` syntax is not required though, correct? I can include it. 

I'll also have the SQL pages up soon that will include the SQL railroad diagrams and more details on the syntax options for users.

Reviewer: It's not _strictly_ required, but it helps make sure that the tenant that gets created has the correct "Capabilities."  If they don't do this, they'll need to run various follow-up commands to make sure their virtual cluster works the same as the one on the primary side.

Author: We have a thread in Slack ongoing re `capabilities`, so I will keep an eye on that and include further detail as needed. For now, added a placeholder sentence. 

I did try to use the `LIKE` syntax on my test clusters... when you specify the `template` are we talking about referring to the VC template from the primary? I received an error `ERROR: retrieving LIKE tenant record: tenant "application" does not exist` when trying to do that. Maybe that's because the app VC on the primary was already shared? 

However, I have a couple of VCs hanging about on the standby, so I tried referring to a diff VC on the standby with `LIKE` when creating a new replication job (and a new standby VC). That worked fine. But, I don't know if that's the point of `LIKE`.

Reviewer: When you create the target cluster using the configuration profile `replication-target` it should create a virtual cluster named `template` in the destination as well that should work for this command on the destination.

The unit of replication is a _virtual cluster_, {% comment %}link to Matt's CV docs{% endcomment %} which is part of the underlying infrastructure in the primary and standby clusters.

In this tutorial, you will connect to:

- The _system interface_ for administration tasks in both clusters, and starting the replication stream from the standby cluster.
- The application _virtual cluster_ on the primary to work with databases, tables, workloads, and so on.

{% include {{ page.version.version }}/physical-replication/reference-links-replication.md %}

## Overview

The high-level steps in this tutorial are:

1. Create and start the primary cluster.
1. Configure and create a user on the primary cluster.
1. Create and start the standby cluster.
1. Configure and create a user on the standby cluster.
1. Verify certificates.
1. Start the replication stream from the standby cluster.

> ## Before you begin

Author: @mdlinville Perhaps, though this is included in the deployment prerequisite along with the certificate generation that you mention in the next comment. I was trying to avoid duplicating too much content we have in that deployment piece. Having said that, if you think it is hard to follow then I could label where node IPs are with a node number rather than actual IPs.

I followed the deploy tutorial with separate cloud VMs, so maybe that did make a difference.

> ## Before you begin

Reviewer 2: Makes sense. I think when I followed the tutorial, I needed to do it with a new cluster, so I read it that way, but of course others will probably be configuring PCR on an existing cluster that already has certificates.

{% include enterprise-feature.md %}

- Two separate CockroachDB clusters (primary and standby) with a minimum of three nodes each using the same CockroachDB {{page.version.version}} version.
    - To set up each cluster, you can follow the [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}). When you start each node in your cluster with the `cockroach start` command, you **must** pass the `--config-profile` flag with a `replication` value. Refer to cluster creation steps for the [primary](#start-the-primary-cluster) and for the [standby](#start-the-standby-cluster) cluster to do this.
- All nodes in each cluster will need access to the Certificate Authority for the other cluster. Refer to [Verify certificates](#step-3-verify-certificates).
- An [{{ site.data.products.enterprise }} license]({% link {{ page.version.version }}/enterprise-licensing.md %}) on the primary **and** standby clusters. You must use the system interface on the primary and standby to enable your {{ site.data.products.enterprise }} license.

> 

Author: Ah yes! I've created an issue for this to follow-up once the initial docs are out.

> ## Step 1. Create the primary cluster

Author: See comment above

### Start the primary cluster

> To enable physical cluster replication, it is necessary to start each node with the `--config-profile` flag and a replication value. This creates a _virtualized cluster_ with a system interface and an application virtual cluster, and sets up all the required configuration for starting a replication stream. {% comment %}Add in links to Matt's overview and descriptions here{% endcomment %}

Author: Done. I like it — this really helps describe the configuration profile better as well. Thanks!

The primary cluster requires the following value:

{% include_cached copy-clipboard.html %}
~~~ shell
--config-profile replication-source
~~~

For example, a `cockroach start` command according to the [prerequisite deployment guide]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}#step-3-start-nodes):

> {% include_cached copy-clipboard.html %}

Author: I did not include this because the initialization happens in the prerequisite tutorial. I added a sentence to make this clear.

~~~ shell
cockroach start \
--certs-dir=certs \
--advertise-addr=<node1 address> \
--join=<node1 address>,<node2 address>,<node3 address> \
--cache=.25 \
--max-sql-memory=.25 \

> --background \

Reviewer 2: ```suggestion
```
Don't start nodes in the background. it's better to start in the foreground so they can see what's happening. See https://github.com/cockroachdb/docs/pull/18033 and [DOC-523] for context.

--config-profile replication-source
~~~

> ### Connect to the primary cluster system interface

Reviewer 2: I like it, and I think it makes sense to have it at the bottom of the topic. I guess we can get further feedback over time.

Connect to your primary cluster's system interface using [`cockroach sql`]({% link {{ page.version.version }}/cockroach-sql.md %}).

1. To connect to the system interface, pass the `options=-ccluster=system` parameter in the URL:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://root@{your IP or hostname}:26257/?options=-ccluster=system&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

    You will find that the prompt will include `system` when you have successfully connected.

    {{site.data.alerts.callout_info}}
    You should only use the system interface for administration. To work with databases, tables, or workloads, use the application virtual cluster{% comment %}({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}){% endcomment %}.
    {{site.data.alerts.end}}

1. Add your cluster organization and {{ site.data.products.enterprise }} license to the cluster:

    {% include_cached copy-clipboard.html %}
    ~~~ sql

>     SET CLUSTER SETTING cluster.organization = 'your organization';

Author: Yes, both are necessary

    ~~~
    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING enterprise.license = 'your enterprise license';
    ~~~

1. Confirm the status of your virtual cluster: {% comment %}Link to SQL ref page here{% endcomment %}

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SHOW VIRTUAL CLUSTERS;
    ~~~

    The output will include the `system` interface and the `application` virtual cluster:

    ~~~
    id |    name     | data_state | service_mode
    ---+-------------+------------+---------------
    1  | system      | ready      | shared
    2  | template    | ready      | none
    3  | application | ready      | shared
    (3 rows)

>     ~~~

Author: Added.

### Create a replication user and password

> The standby cluster requires a user profile on the primary cluster to connect to.

Author: Done!

> 

Reviewer 2: I think at the same time you should have them create a DB Console user on the primary cluster. That way, when the cut over to the standby, they can immediately use that identity to log in and verify that all the databases they expect to see in the application cluster, are present.

Author: I think this was captured in one of the steps where I have now created a users.

1. From the primary's system interface SQL shell, create a user and password:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE USER {your username} WITH LOGIN PASSWORD '{your password}';
    ~~~

1. Grant the [`REPLICATION` system privilege]({% link {{ page.version.version }}/security-reference/authorization.md %}#supported-privileges) to your user: {% comment %}Link to detail on replication system privilege{% endcomment %}

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    GRANT SYSTEM REPLICATION TO {your username};
    ~~~

    If you need to change the password later, refer to [`ALTER USER`]({% link {{ page.version.version }}/alter-user.md %}).

### Connect to the primary virtual cluster (optional)

> 1. If you would like to connect to the primary's application virtual cluster, open a new terminal window, and use the `ccluster=application` parameter:

Author: Added in the `cockroach workload init` + `run` commands, plus gave the detail on the connection string to the primary application to do this.

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://root@{your IP or hostname}:26257/?options=-ccluster=application&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

    Your prompt will include `application` when you have successfully connected.

    {{site.data.alerts.callout_info}}

>     You can connect to the [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) to observe activity on the primary cluster. If you did not already [create a user in the previous step](#create-a-replication-user-and-password), create one to access the DB Console. Open a web browser at `https://{your IP or hostname}:8080/` and enter your credentials.

Author: Had them create a user here. Made this a fully fledged sub step

    {{site.data.alerts.end}}

## Step 2. Create the standby cluster

### Start the standby cluster

Similarly to the primary cluster, it is necessary to start each node with the `--config-profile` flag and a `replication-target` for the standby cluster. This creates a _virtualized cluster_ with a system interface, and sets up all the required configuration for starting a replication stream.

To start the standby cluster, run:

{% include_cached copy-clipboard.html %}
~~~ shell
--config-profile replication-target
~~~

For example, a `cockroach start` command according to the [prerequisite deployment guide]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}#step-3-start-nodes):

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach start \
--certs-dir=certs \
--advertise-addr=<node1 address> \
--join=<node1 address>,<node2 address>,<node3 address> \
--cache=.25 \
--max-sql-memory=.25 \
--background \
--config-profile replication-target
~~~

### Connect to the standby cluster system interface

> Connect to your standby cluster's system interface using [`cockroach sql`]({% link {{ page.version.version }}/cockroach-sql.md %}).

Author: I added a sentence for this.

1. Add your cluster organization and {{ site.data.products.enterprise }} license to the cluster:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING cluster.organization = 'your organization';
    ~~~
    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SET CLUSTER SETTING enterprise.license = 'your enterprise license';
    ~~~

1. Confirm the status of your virtual cluster:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SHOW VIRTUAL CLUSTERS;
    ~~~

    The output will show the `system` interface, but no `application` virtual cluster:

    ~~~
    id |   name   | data_state | service_mode
    ---+----------+------------+---------------
    1  | system   | ready      | shared
    2  | template | ready      | none
    (2 rows)
    ~~~

    {% include {{ page.version.version }}/physical-replication/template-description.md %}

### Create a user for the standby cluster

If you would like to access the DB Console to observe your replication, you will need to create a user:

1. Create a user:

    {% include_cached copy-clipboard.html %}
    ~~~ sql

>     CREATE USER {your username} WITH LOGIN PASSWORD {'your password'};

Author: Changed.

    ~~~

1. To observe the replication activity, your user will need [`admin`]({% link {{ page.version.version }}/security-reference/authorization.md %}#admin-role) privileges:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    GRANT admin TO {your username};
    ~~~

    Open the DB Console in your web browser: `https://{your IP or hostname}:8080/`, where you will be prompted for these credentials. Refer to [Physical Cluster Replication Monitoring]({% link {{ page.version.version }}/physical-cluster-replication-monitoring.md %}) for more detail on tracking relevant metrics for your replication stream.

> ## Step 3. Verify certificates

Reviewer 2: ```suggestion
## Step 3. Copy certificates
```
You aren't verifying anything here. You're verifying that your certificates are set up correctly in the next section by setting up replication. It won't work if certificates aren't correct. Ask me how I know. LOL

Author: Done!

At this point, you have a primary and a standby cluster running. The next step is to verify that the clusters can communicate. Depending on how you manage certificates, you must ensure that all nodes on the primary and the standby cluster have access to the certificate of the other cluster.

For example, if you followed the [Deploy CockroachDB]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}) prerequisite, you need to add the `ca.crt` from the primary cluster to the `certs` directory on all the nodes in the standby cluster.

1. Name the `ca.crt` from the primary cluster to a new name on the standby cluster. For example, `ca_primary.crt`.

> 1. Place `ca_primary.crt` in the `certs` directory of the standby cluster nodes.

Reviewer 2: ```suggestion
1. Securely copy `ca_primary.crt` to the `certs` directory of the standby cluster nodes.
```

Author: Done!

You need to add the `ca.crt` from the standby cluster to the `certs` directory on all the nodes in the primary cluster.

1. Name the `ca.crt` from the standby cluster to a new name on the primary cluster. For example, `ca_standby.crt`.

> 1. Place `ca_standby.crt` in the `certs` directory of the primary cluster nodes.

Reviewer 2: ```suggestion
1. Securely copy `ca_standby.crt` to the `certs` directory of the primary cluster nodes.
```

Author: Done!

    - The path to the primary node's certificate on the standby cluster.

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE VIRTUAL CLUSTER standbyapplication LIKE template
    FROM REPLICATION OF application
    ON 'postgresql://{replication user}:{password}@{node IP or hostname}:26257/?options=-ccluster=system&sslmode=verify-full&sslrootcert=certs/{primary cert}.crt';
    ~~~

    {% include {{ page.version.version }}/physical-replication/like-description.md %}

    Once the standby cluster has made a connection to the primary cluster, the standby will pull the topology of the primary cluster and will distribute the replication work across all nodes in the primary and standby.

1. To manage the replication stream, you can pause and resume the replication stream as well as show the current details for the job:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER standbyapplication PAUSE REPLICATION;
    ~~~

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    ALTER VIRTUAL CLUSTER standbyapplication RESUME REPLICATION;
    ~~~

    {% include_cached copy-clipboard.html %}
    ~~~ sql

>     SHOW VIRTUAL CLUSTER standbyapplication WITH REPLICATION STATUS;

Reviewer 2: It might be worth showing the concise output from `SHOW VIRTUAL CLUSTERS;` here and point out the `replicating` status. I found it easier to read.

Author: Added this right beneath the replication stream creation command.

---
title: Set Up Physical Cluster Replication
summary: Follow a tutorial to set up physical replication between a primary and standby cluster.
toc: true
docs_area: manage
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

> {% include_cached new-in.html version="v23.2" %} In this tutorial, you will set up physical cluster replication between a primary cluster and standby cluster. The primary is _active_, serving application traffic. The standby is _passive_, accepting updates from the primary cluster. The replication stream will send changes from the primary to the standby.

Reviewer 2: ```suggestion
{% include_cached new-in.html version="v23.2" %} In this tutorial, you will set up physical cluster replication between a primary cluster and standby cluster. The primary cluster is _active_, serving application traffic. The standby cluster is _passive_, accepting updates from the primary cluster. The replication stream will send changes from the primary to the standby.
```

Author: Done!

The unit of replication is a _virtual cluster_, {% comment %}link to Matt's CV docs{% endcomment %} which is part of the underlying infrastructure in the primary and standby clusters.

In this tutorial, you will connect to:

- The _system interface_ for administration tasks in both clusters, and starting the replication stream from the standby cluster.
- The application _virtual cluster_ on the primary to work with databases, tables, workloads, and so on.

{% include {{ page.version.version }}/physical-replication/reference-links-replication.md %}

## Overview

The high-level steps in this tutorial are:

1. Create and start the primary cluster.
1. Configure and create a user on the primary cluster.
1. Create and start the standby cluster.
1. Configure and create a user on the standby cluster.
1. Securely copy certificates.
1. Start the replication stream from the standby cluster.

## Before you begin

{% include enterprise-feature.md %}

> - Two separate CockroachDB clusters (primary and standby) with a minimum of three nodes each using the same CockroachDB {{page.version.version}} version.

Reviewer 2: ```suggestion
- Two separate CockroachDB clusters (primary and standby) with a minimum of three nodes each, and each using the same CockroachDB {{page.version.version}} version.
```

Author: Done!

>     - To set up each cluster, you can follow the [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}). When you start each node in your cluster with the `cockroach start` command, you **must** pass the `--config-profile` flag with a `replication` value. Refer to cluster creation steps for the [primary](#start-the-primary-cluster) and for the [standby](#start-the-standby-cluster) cluster to do this.

Reviewer 2: ```suggestion
    - To set up each cluster, you can follow [Deploy CockroachDB on Premises]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}). When you start each node in your cluster with the `cockroach start` command, you **must** pass the `--config-profile` flag with a `replication` value. Refer to cluster creation steps for the [primary cluster](#start-the-primary-cluster) and for the [standby cluster](#start-the-standby-cluster) cluster for details.
```

Author: Done!

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach start \
--certs-dir=certs \
--advertise-addr=<node1 address> \
--join=<node1 address>,<node2 address>,<node3 address> \
--cache=.25 \
--max-sql-memory=.25 \
--background \
--config-profile replication-source
~~~

Ensure that you follow the [prerequisite deployment guide]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}#step-4-initialize-the-cluster) to initialize your cluster before continuing to set up physical cluster replication.

### Connect to the primary cluster system interface

Connect to your primary cluster's system interface using [`cockroach sql`]({% link {{ page.version.version }}/cockroach-sql.md %}).

1. To connect to the system interface, pass the `options=-ccluster=system` parameter in the URL:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://root@{node IP or hostname}:26257/?options=-ccluster=system&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

>     You will find that the prompt will include `system` when you have successfully connected.

Reviewer 2: ```suggestion
    The prompt will include `system` when you are connected to the system interface.
```

Author: Done!

    {{site.data.alerts.callout_info}}

>     You should only use the system interface for administration. To work with databases, tables, or workloads, use the application virtual cluster{% comment %}({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}){% endcomment %}.

Reviewer 2: ```suggestion
    You should only connect to the system interface for cluster administration. To work with databases, tables, or workloads, you connect to the application virtual cluster{% comment %}({% link {{ page.version.version }}/physical-cluster-replication-overview.md %}){% endcomment %}.
```

Author: Nice, done!

1. From the primary's system interface SQL shell, create a user and password:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE USER {your username} WITH PASSWORD '{your password}';
    ~~~

1. Grant the [`REPLICATION` system privilege]({% link {{ page.version.version }}/security-reference/authorization.md %}#supported-privileges) to your user: {% comment %}Link to detail on replication system privilege{% endcomment %}

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    GRANT SYSTEM REPLICATION TO {your username};
    ~~~

    If you need to change the password later, refer to [`ALTER USER`]({% link {{ page.version.version }}/alter-user.md %}).

### Connect to the primary virtual cluster (optional)

1. If you would like to run a sample workload on the primary's application virtual cluster, open a new terminal window and use [`cockroach workload`]({% link {{ page.version.version }}/cockroach-workload.md %}) to run the workload.

    For example, to initiate the [`movr`]({% link {{ page.version.version }}/movr.md %}) workload:

    {% include_cached copy-clipboard.html %}
    ~~~ shell

>     cockroach workload init movr "postgresql://root@{node IP or hostname}:{26257}/?options=-ccluster=application&sslmode=verify-full&sslrootcert=certs/ca.crt&sslcert=certs/client.root.crt&sslkey=certs/client.root.key"
>     ~~~
> 

Reviewer 2: ```suggestion
    cockroach workload init movr "postgresql://root@{node_advertise_address}:{node_advertise_port}/?options=-ccluster=application&sslmode=verify-full&sslrootcert=certs/ca.crt&sslcert=certs/client.root.crt&sslkey=certs/client.root.key"
    ~~~

Replace `{node_advertise_address}` and `{node_advertise_port}` with a node's `--advertise-address` IP address or hostname and port.
```
WDYT? I think including a static port might cause some confusion, but I don't feel super strongly about it.

Author: Added and linked to the `cockroach start` flags section for `advertise-addr`

    {% include {{ page.version.version }}/connect/cockroach-workload-parameters.md %} As a result, for the example in this tutorial, you will need:
    - `options=-ccluster=application`
    - `sslmode=verify-full`
    - `sslrootcert={path}/certs/ca.crt`: the path to the CA certifcate.
    - `sslcert={path}/certs/client.root.crt`: the path to the client certificate.
    - `sslkey={path}/certs/client.root.key`: the path to the client private key.

    For additional detail on the standard CockroachDB connection parameters, refer to [Client Connection Parameters]({% link {{ page.version.version }}/connection-parameters.md %}#connect-using-a-url).

1. Run the `movr` workload for a set duration using the same connection string:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach workload run movr --duration=5m "postgresql://root@{node IP or hostname}:{26257}/?options=-ccluster=application&sslmode=verify-full&sslrootcert=certs/ca.crt&sslcert=certs/client.root.crt&sslkey=certs/client.root.key"
    ~~~

> 1. To connect to the primary's application virtual cluster, use the `ccluster=application` parameter:

Reviewer 2: ```suggestion
1. To connect to the primary cluster's application virtual cluster, use the `ccluster=application` parameter:
```
I think it might be worthwhile to always refer to "primary cluster" and standby cluster" rather than implying them.

Author: I have changed up where I have found these

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://root@{node IP or hostname}:26257/?options=-ccluster=application&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

>     Your prompt will include `application` when you have successfully connected.

Reviewer 2: ```suggestion
    The prompt will include `application` when you are connected to the application virtual cluster.
```

Author: Done!

1. Create a user for your primary's application cluster:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE USER {your username} WITH PASSWORD '{your password}';
    ~~~

1. You can connect to the [DB Console]({% link {{ page.version.version }}/ui-overview.md %}) with this user to observe activity on the primary cluster. Open a web browser at `https://{node IP or hostname}:8080/` and enter your credentials.

## Step 2. Create the standby cluster

### Start the standby cluster

> Similarly to the primary cluster, it is necessary to start each node with the `--config-profile` flag and a `replication-target` for the standby cluster. This creates a _virtualized cluster_ with a system interface, and sets up all the required configuration for starting a replication stream.

Reviewer 2: ```suggestion
Similarly to the primary cluster,each node on the standby cluster must be started with the `--config-profile` flag set to `replication-target`. This creates a _virtualized cluster_ with a system interface and an application cluster, and sets up all the required configuration for starting a replication stream.
```

Author: Nice, I like this, changed!

Reviewer 2: For followup, I don't think that the standby cluster actually gets an auto-created virtual cluster. You specifically walk them through creating it.

{% include_cached copy-clipboard.html %}
~~~ shell
cockroach start \
--certs-dir=certs \
--advertise-addr=<node1 address> \
--join=<node1 address>,<node2 address>,<node3 address> \
--cache=.25 \
--max-sql-memory=.25 \
--background \
--config-profile replication-target
~~~

Ensure that you follow the [prerequisite deployment guide]({% link {{ page.version.version }}/deploy-cockroachdb-on-premises.md %}#step-4-initialize-the-cluster) to initialize your cluster before continuing to set up physical cluster replication.

### Connect to the standby cluster system interface

Connect to your standby cluster's system interface using [`cockroach sql`]({% link {{ page.version.version }}/cockroach-sql.md %}).

1. To connect to the system interface, pass the `options=-ccluster=system` parameter in the URL:

    {% include_cached copy-clipboard.html %}
    ~~~ shell
    cockroach sql --url \
    "postgresql://root@{node IP or hostname}:26257/?options=-ccluster=system&sslmode=verify-full" \
    --certs-dir "certs"
    ~~~

>     You will find that the prompt will include `system` when you have successfully connected.

Reviewer 2: ```suggestion
    The prompt will include `system` when you are connected to the system interface.
```

Author: Done!

    ---+----------+------------+---------------
    1  | system   | ready      | shared
    2  | template | ready      | none
    (2 rows)
    ~~~

    {% include {{ page.version.version }}/physical-replication/template-description.md %}

### Create a user for the standby cluster

If you would like to access the DB Console to observe your replication, you will need to create a user:

1. Create a user:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    CREATE USER {your username} WITH LOGIN PASSWORD {'your password'};
    ~~~

1. To observe the replication activity, your user will need [`admin`]({% link {{ page.version.version }}/security-reference/authorization.md %}#admin-role) privileges:

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    GRANT admin TO {your username};
    ~~~

    Open the DB Console in your web browser: `https://{node IP or hostname}:8080/`, where you will be prompted for these credentials. Refer to [Physical Cluster Replication Monitoring]({% link {{ page.version.version }}/physical-cluster-replication-monitoring.md %}) for more detail on tracking relevant metrics for your replication stream.

> ## Step 3. Copy certificates

Reviewer 2: I think some links to here still call this section "Verify certificates".

Author: Found, and changed

> At this point, you have a primary and a standby cluster running. The next step is to verify that the clusters can communicate. Depending on how you manage certificates, you must ensure that all nodes on the primary and the standby cluster have access to the certificate of the other cluster.

Reviewer 2: ```suggestion
At this point, the primary and standby clusters are both running. The next step allows the standby cluster to connect to the primary cluster and begin ingesting its data. Depending on how you manage certificates, you must ensure that all nodes on the primary and the standby cluster have access to the certificate of the other cluster.
```

Author: Edited, nice!

    {% include_cached copy-clipboard.html %}
    ~~~ sql
    SHOW VIRTUAL CLUSTER standbyapplication WITH REPLICATION STATUS;
    ~~~

    {% comment %}update output here{% endcomment %}

    {% include_cached copy-clipboard.html %}
    ~~~
    id |        name        |     data_state     | service_mode | source_tenant_name |                                                     source_cluster_uri                                               | replication_job_id |        replicated_time        |         retained_time         | cutover_time
    ---+--------------------+--------------------+--------------+--------------------+----------------------------------------------------------------------------------------------------------------------+--------------------+-------------------------------+-------------------------------+---------------
    3  | standbyapplication | replicating        | none         | application        | postgresql://{user}:{password}@{hostname}:26257/?options=-ccluster%3Dsystem&sslmode=verify-full&sslrootcert=redacted | 899090689449132033 | 2023-09-11 22:29:35.085548+00 | 2023-09-11 16:51:43.612846+00 |     NULL
    (1 row)
    ~~~

## Connection reference

> The following table outlines the connection strings you will need for this setup tutorial.

Reviewer 2: ```suggestion
This table outlines the connection strings you will need for this setup tutorial.
```

Author: Done!

### Connect to the primary virtual cluster (optional)

1. If you would like to run a sample workload on the primary's application virtual cluster, open a new terminal window and use [`cockroach workload`]({% link {{ page.version.version }}/cockroach-workload.md %}) to run the workload.

    For example, to initiate the [`movr`]({% link {{ page.version.version }}/movr.md %}) workload:

    {% include_cached copy-clipboard.html %}
    ~~~ shell

>     cockroach workload init movr "postgresql://root@{node IP or hostname}:{26257}/?options=-ccluster=application&sslmode=verify-full&sslrootcert=certs/ca.crt&sslcert=certs/client.root.crt&sslkey=certs/client.root.key"

Reviewer 2: I don't think `cockroach workload` accepts a connection string, although it doesnt accept `--secure`. Can you verify? When I tried to pass it a connection string, it did not work. If you are running on the default SQL port, `cockroach workload` seems to hit that port and do the right thing. Otherwise I think you can give it a port directly. https://www.cockroachlabs.com/docs/stable/cockroach-workload

Author: Ah ok, so I was working directly on the cockroach SQL port, so it worked just fine with the connection string. 

There is the `port=` param which seems to work. How did you get your `cockroach workload` to run? I don't know which flag would work to pass in a diff flag?

Reviewer 2: I just ran it locally on one of the nodes (the one that was listening on the default port, as it happens). I did see that you have added src/current/_includes/v23.2/connect/cockroach-workload-parameters.md and I didn't try doing it using a postgresql:// address or didn't know that was a possibility. So maybe this is fine how you have it, and in fact better than mine, which was working in spite of me rather than because of me.
